### PR TITLE
Fix error response handling in Orchestrator Zone CRUD apis

### PR DESCRIPTION
**Pull request by babamishra9**
- Return error response in application/json content type
- Format with error message in "message" field as expected by Orchestrator Controller.
- Move all exception handlers into this inner class.
- Added missing empty/space check for name parameter in service

### CODE REVIEW

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java [hunk 1]
```diff
@@ -255,10 +266,11 @@ void testDeleteZone_Forbidden() throws Exception {
 
     @Test
     void testUpdateZone_MethodNotImplemented() throws Exception {
-        performMockMvcCallAndAssertError(put("/orchestrator/zones").contentType(APPLICATION_JSON).content(
+        performMockMvcCallAndAssertError(put("/orchestrator/zones").contentType(APPLICATION_JSON_VALUE).content(
```
Comment (irinaepshteyn) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java:269]: I believe this change is not needed.  Request content type should be set to APPLICATION_JSON

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java [hunk 2]
```diff
@@ -221,11 +222,20 @@ public void createZone(OrchestratorZoneRequest zoneRequest) throws OrchestratorZ
         }
     }
 
+    private String getName(OrchestratorZoneRequest zoneRequest) throws OrchestratorZoneServiceException {
```
Comment (irinaepshteyn) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java:225]: Purpose of this PR and corresponding user story is to fix format of the error message to valid json.  Adding name validation is not part of this story.  Also, this change is not covered by any tests. These changes should be done in the context of input validation story.

Remove changes to OrchestratorZoneService class.

Comment (irinaepshteyn) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java:225]: Consider refactoring OrchestratorZoneService deleteZone() method.  To be consistent with the rest of the public methods, it should not return ResponseEntity<?>; it should return void instead.  Error handling and ResponseEntity should be moved into OrchestratorZoneController.

In OrchestratorZoneService refactor deleteZone() :
```
    public void deleteZone(String zoneName) {
        IdentityZone previous = IdentityZoneHolder.get();
        try {
            ...
            if (publisher != null && zone != null) {
                ...
                return;
            } else {
                throw new Exception(...);
            }
        } finally {
            IdentityZoneHolder.set(previous);
        }
    }
```

In OrchestratorZoneController, refactor deleteZone():
```
    @DeleteMapping
    public ResponseEntity<?> deleteZone(@NotBlank(message = MANDATORY_VALIDATION_MESSAGE) @RequestParam String name) {
        return ResponseEntity.accepted().build();
    }
```
That refactoring fits under "fix error response handling..."

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
@babamishra9 , done reviewing.  Please address my comments

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 3]
```diff
@@ -70,54 +73,44 @@ public ResponseEntity<?> updateZone(@RequestBody OrchestratorZoneRequest zoneReq
...
-    @ExceptionHandler(OrchestratorZoneServiceException.class)
-    public ResponseEntity<String> handleOrchestratorZoneServiceException(OrchestratorZoneServiceException e) {
-        return new ResponseEntity<>("{\"message\", \""+ e.getMessage() +"\" }", BAD_REQUEST);
-    }
-
-    @ExceptionHandler(Exception.class)
-    public ResponseEntity<String> handleException(Exception e) {
-        logger.error(e.getClass() + ": " + e.getMessage(), e);
-        return new ResponseEntity<>("{\"message\", \""+ e.getMessage() +"\" }", INTERNAL_SERVER_ERROR);
+    @RestControllerAdvice
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java:76]: There are few advantages on using RestControllerAdvice, as  for common statuscode as bad request , multiple exception can be handled as part of one handler method. for example.
 @ExceptionHandler(value = { HttpMessageNotReadableException.class, MethodArgumentNotValidException.class,
                                    MissingServletRequestParameterException.class, ConstraintViolationException.class,
                                    ZoneAlreadyExistsException.class, OrchestratorZoneServiceException.class})

In addition, we can just understand it as:
@RestControler = @Controller + @ResponseBody
@RestControllerAdvice = @ControllerAdvice + @ResponseBody.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java (hunk 1 above)
Comment (babamishra9) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/OrchestratorZoneControllerMockMvcTests.java:269]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 4]
```diff
@@ -70,54 +73,44 @@ public ResponseEntity<?> updateZone(@RequestBody OrchestratorZoneRequest zoneReq
...
+    private class OrchestratorZoneExceptionHandler
+    {
+        @ExceptionHandler(value = { HttpMessageNotReadableException.class, MethodArgumentNotValidException.class,
+                                    MissingServletRequestParameterException.class, ConstraintViolationException.class,
+                                    ZoneAlreadyExistsException.class, OrchestratorZoneServiceException.class})
+        @ResponseStatus(value = BAD_REQUEST)
+        public OrchestratorErrorResponse badRequest(Exception ex)
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 5]
```diff
@@ -50,14 +53,14 @@ public OrchestratorZoneController(OrchestratorZoneService zoneService) {
 
     @GetMapping
     public ResponseEntity<OrchestratorZoneResponse> getZone(@NotBlank(message = MANDATORY_VALIDATION_MESSAGE) @RequestParam String name) {
-        return new ResponseEntity<>(zoneService.getZoneDetails(name), HttpStatus.OK);
+        return new ResponseEntity<>(zoneService.getZoneDetails(name),OK);
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java [hunk 6]
```diff
@@ -50,14 +53,14 @@ public OrchestratorZoneController(OrchestratorZoneService zoneService) {
...
     }
 
     @PostMapping
     public ResponseEntity<?> createOrchestratorZone(@RequestBody @Valid OrchestratorZoneRequest orchestratorZoneRequest )
         throws OrchestratorZoneServiceException, IOException {
         zoneService.createZone(orchestratorZoneRequest);
-        return new ResponseEntity<>("", HttpStatus.ACCEPTED);
+        return new ResponseEntity<>(ACCEPTED);
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/OrchestratorErrorResponse.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
+@JsonInclude(Include.NON_NULL)
+public class OrchestratorErrorResponse {
+    private String message;
+}
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/OrchestratorErrorResponse.java:11]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java (hunk 2 above)
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java]: Done. Thanks!

Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java]: > Purpose of this PR and corresponding user story is to fix format of the error message to valid json. Adding name validation is not part of this story. Also, this change is not covered by any tests. These changes should be done in the context of input validation story.
> 
> Remove changes to OrchestratorZoneService class.

@babamishra9 : Done. Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneController.java]: As discussed, removed RestControllerAdvice and replacing them with ExceptionHandlers.

Thanks!

**Review by irinaepshteyn — APPROVED** [review]
Looks good.

**Pull request merged**